lightning/src/ln/channelmanager.rs

```diff
@@ -9050,7 +9104,38 @@ where
 		let hmac = payment_id.hmac_for_offer_payment(nonce, expanded_key);
 		let context = OffersContext::OutboundPayment { payment_id, nonce, hmac: Some(hmac) };
 		let reply_paths = self.create_blinded_paths(context)
-			.map_err(|_| Bolt12SemanticError::MissingPaths)?;
+			.map_err(|()| Bolt12RequestError::BlindedPathCreationFailed)?;
+
+		let total_liquidity: u64 = {
+			let per_peer_state = &self.per_peer_state.read().unwrap();
+			let mut sum: u64 = 0;
+
+			per_peer_state.iter().for_each(|(_, peer_state_mutex)| {
+				let peer_state_lock = peer_state_mutex.lock().unwrap();
+				let peer_state = &*peer_state_lock;
+
+				sum += peer_state.channel_by_id
+					.iter()
+					.map(|(_, phase)| phase.context().get_available_balances(&self.fee_estimator).outbound_capacity_msat)
+					.sum::<u64>();
+			});
+			sum
+		};
+
+		let requested_amount_msats = match invoice_request.amount_msats() {
+			Some(amount_msats) => Some(amount_msats),
+			None => match offer.amount() {
+				Some(Amount::Bitcoin { amount_msats }) => Some(amount_msats),
+				_ => None,
+			},
+		};
+
+		if let Some(amount) = requested_amount_msats {
+			if amount > total_liquidity {
+				log_error!(self.logger, "Insufficient liquidity for payment with payment id: {}", payment_id);
+				return Err(Bolt12RequestError::InsufficientLiquidity);
+			}
+		}
```

Comment:
@tnull Should we consider outstanding invoice requests / refunds?

Reply:
Not sure I understand the point. For BOLT11, if we initiate a bunch of payments, each subsequent one will have less liquidity to consider, IIUC. But for BOLT12, we can similarly initiate more than one payment but each subsequent one will consider the same liquidity unless an invoice is received for a prior one.

Reply:
Sorry for the late reply! Since outstanding invoice requests and refunds can tie up liquidity, do you think we should check for any outstanding invoices and see how they affect available liquidity before moving forward with the payment? And should I address it in this PR? I want to make sure to cover all our bases.

Reply:
So they can potentially _use_ liquidity if an invoice is received in response. But they don't _tie up_ liquidity since a payment hasn't been made yet. That said, we can consider a request for payment as an intention to tie up liquidity in the hopefully near future. I'd lean towards including it but will defer to @tnull.

FWIW, there's also some edge cases to consider. If we have a fiat-denominated offer and the users *doesn't* give an amount, how much liquidity should we considered unavailable?